lib/ymodel/relatable.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require 'active_support/inflector'
+require 'ymodel/errors'
+
+module YModel
+  # This module contains YModel logic for managing relations. Some caveats:
+  #   - combining the 'as' and 'foreign_key' options won't work.
+  #   - the default foreign_key won't contain namespaces.
+  #   - Seems to be the most smelly piece of code in YModel.
+  module Relatable
+    def belongs_to(model, _options = {})
+      define_method(model) do
+        foreign_key = model.to_s.singularize + '_id'
+        relation_class.find(instance_variable_get(foreign_key))
+      end
+    end
+
+    # These cop is disabled because I'm just copying ActiveRecords interface
+    # rubocop:disable Naming/PredicateName,
+    # rubocop:disable Naming/UncommunicativeMethodParamName
+    def has_many(model, class_name: nil, as: nil, foreign_key: nil, **_options)
+      raise_options_error if as && foreign_key
+
+      foreign_key   ||= default_foreign_key
+      relation_class  = model_class(class_name || model)
+      klass_name      = klass_name
+      define_method(model) do
+        if as
+          relation_class.where("#{as}_id" => id,
+                               "#{as}_type" => klass_name)
+        else
+          relation_class.where(foreign_key.to_sym => id)
+        end
+      end
+    end
+
+    def has_one(model, class_name: nil, as: nil, foreign_key: nil, **_options)
+      raise_options_error if as && foreign_key
+
+      foreign_key   ||= default_foreign_key
+      relation_class  = model_class(class_name || model)
+      klass_name      = klass_name
+
+      define_method(model) do
+        if as
+          relation_class.find_by("#{as}_id" => id,
+                                 "#{as}_type" => klass_name)
+        else
+          relation_class.find_by(foreign_key => id)
+        end
+      end
+    end
+    # rubocop:enable Naming/PredicateName,
+    # rubocop:enable Naming/UncommunicativeMethodParamName
+
+    private
+
+    def default_foreign_key
+      klass_name.underscore
+        .split('/')
+        .last
```

Comment:
I'm not sure why RuboCop/Hound is giving you these messages. Indenting by two spaces is consistent with the rest of our code, so I wouldn't change this. My guess is moving `.underscore` will silence the message:

```ruby
klass_name
  .split('/)
  .last
  .underscore
```

... though I think this chain fits within 80 characters, so it could be:

```ruby
klass_name.split('/').last.underscore
````

(`underscore` last should be functionally the same, but means fewer chars have to be considered).

Reply:
Ik think its quite important to split method calls to different lines once you call more then two. This makes it really easy to detect code that doesn't comply with [the law of demeter](https://en.wikipedia.org/wiki/Law_of_Demeter).

I think we should look into improving the config of Hound and Rubocop. Its my view the rules are a bit to strict right now and even worse, they are conflicting. If we'd make the linter a bit more lenient I'm pretty sure I would run into less issues, and thus produce more output, without the code quality suffering to much. It would also become a whole lot less frustrating to deal with it for me.

I'll try and lobby for some time to improve the linting with Chael.